### DISCUSSION TITLE
ESP32-P4 released, support?

### DISCUSSION THREAD

**Discussion by Yelgurk** [lovyan03/LovyanGFX#555] [Device Configurations]
Good afternoon! ESP32-P4 chip has been released.
- https://www.espressif.com/en/products/socs/esp32-p4
- https://www.youtube.com/watch?v=GTHHDZeIzcA

It is still difficult to buy it on the free market, but already some companies seem to be able to buy the first revisions and build modules based on these chips.
- https://www.wireless-tag.com/product-item-61.html
- https://img01.71360.com/w3/77q677/20240409/025a34d13f1a027cb65d0cd9e1f52467.pdf?dl=1&dlf=WT01P4C6-S1%E8%A7%84%E6%A0%BC%E4%B9%A6V1.1.pdf

**Comment by Yelgurk** [comment]
Some additional info, esp32-p4 datasheet: https://www.erlendervik.no/ESP32-C5%20Beta_ESP32-P4_ESP8686_ESP32-C3FH4X/ESP32_P4_Chip_Datasheet_V0.1_PRELIMINARY_EN.pdf

**Comment by tobozo** [comment]
hi

esp32-p4 support will be considered as soon as v5.3 hits the support matrix

![image](https://github.com/lovyan03/LovyanGFX/assets/1893754/2a059863-f1ae-4dea-b66c-3616a26610fb)

![image](https://github.com/lovyan03/LovyanGFX/assets/1893754/d2467de2-55e3-4fa3-b044-fcabeef3926b)

**Reply by SlyRebula** [reply]
any news on esp32 p4 support ?

**Reply by tobozo** [reply]
now that they use AI, google search results are so mediocre they make aliexpress search results look pertinent :rofl:

that's the only p4-based product I could find (apart from the 120$ ev-kit from espressif), and it's out of stock :

https://www.waveshare.com/esp32-p4-nano.htm 

let's hope for some p4-based product announcements after the Chinese new year 🐍

**Reply by SlyRebula** [reply]
Thank you for willing to buy it to develop support! You might able to buy from this site for cheaper i think only diffrence is it has no 32 or 16 mb psram: https://www.olimex.com/Products/IoT/ESP32-P4/

**Reply by tobozo** [reply]
Thanks, I've ordered the olimex kit :+1: 

The first objective for LovyanGFX will be to validate i2c/spi legacy bus support for existing panels with ESP32-P4. 

MIPI/QSPI support for LovyanGFX isn't ready yet, but I'm wondering how many (if not all) of those raspberry pi panels it 'll bring into LovyanGFX scope

**Reply by tobozo** [reply]
found another P4-based board, from the CYD family https://www.aliexpress.com/item/1005008328088576.html

**Reply by tobozo** [reply]
@SlyRebula just received the olimex board, and it has 32mb of psram, Olimex probably forgot to add that detail to the product description, it's visible on the back of the board anyway.

![image](https://github.com/user-attachments/assets/3db56eec-8db1-4faf-90a0-9c90a39764ea)

I don't have the skills to do all these tasks, but here's the what I have identified so far :

- Add `#if defined CONFIG_IDF_TARGET_ESP32P4`  conditional macros wherever include errors occur
- Migrate SPI/I2C : `periph_module_{enable,disable,reset}()` functions are all [deprecated with ESP32P4](https://github.com/espressif/esp-idf/blob/0f0068fff3ab159f082133aadfa9baf4fc0c7b8d/components/esp_hw_support/include/esp_private/periph_ctrl.h#L69-L88), recommended implementation is to "use RCC lock macros", but I have no idea what this means.
- Fix DMA : according to esp_lcd library, ESP32P4 should use `dma2d_reg.h`, `dma2d_channel.h` and `dma2d_struct.h` with MIPI-DSI whereas LovyanGFX uses a generic SPI DMA channel

**Comment by tobozo** [comment]
followup for the [Olimex ESP32-P4-DevKit](https://www.olimex.com/Products/IoT/ESP32-P4/):

SSD1306: although there are some warnings, I2C seems to work

ILI9341: 

- some warnings are emitted at compilation:

```log
~/.arduino15/packages/esp32-3.2.0/tools/esp32-arduino-libs/idf-release_v5.4-2f7dcd86-v1/esp32p4/include/esp_hw_support/include/esp_private/periph_ctrl.h:87:64: warning: invalid suffix on literal; C++11 requires a space between literal and string macro [-Wliteral-suffix]
   87 | #define __PERIPH_CTRL_DEPRECATE_ATTR __attribute__((deprecated("This function is not functional on "CONFIG_IDF_TARGET)))
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp: In function 'cpp::bitwizeshift::result<void, lgfx::v1::error_t> lgfx::v1::spi::init(int, int, int, int, int)':
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp:593:7: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  593 |       *reg(SPI_USER_REG(spi_port)) = SPI_USR_MOSI | SPI_USR_MISO | SPI_DOUTDIN;  // need SD card access (full duplex setting)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: source object is likely at address zero
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp: In function 'cpp::bitwizeshift::result<void, lgfx::v1::error_t> lgfx::v1::spi::initQuad(int, int, int, int, int, int, int)':
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp:673:7: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  673 |       *reg(SPI_USER_REG(spi_port)) = SPI_USR_MOSI | SPI_USR_MISO | SPI_DOUTDIN;  // need SD card access (full duplex setting)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: source object is likely at address zero
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/Bus_SPI.cpp: In member function 'virtual void lgfx::v1::Bus_SPI::beginTransaction()':
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/Bus_SPI.cpp:278:5: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  278 |     *reg(SPI_PIN_REG(spi_port)) = pin;
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: source object is likely at address zero
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp: In function 'void lgfx::v1::spi::beginTransaction(int, uint32_t, int)':
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp:763:7: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  763 |       *reg(SPI_USER_REG(spi_port)) = user;
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: source object is likely at address zero
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp:767:7: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  767 |       *reg(SPI_MISC_REG( spi_port)) = pin;
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: source object is likely at address zero
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp:769:7: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  769 |       *reg(SPI_CLOCK_REG(spi_port)) = clkdiv;
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: source object is likely at address zero
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp: In function 'void lgfx::v1::spi::writeBytes(int, const uint8_t*, size_t)':
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp:799:7: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  799 |       *reg(SPI_MOSI_DLEN_REG(spi_port)) = (len << 3) - 1;
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: note: source object is likely at address zero
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp: In function 'void lgfx::v1::spi::readBytes(int, uint8_t*, size_t)':
~/Arduino/libraries/LovyanGFX/src/lgfx/v1/platforms/esp32/common.cpp:810:7: warning: array subscript 0 is outside array bounds of 'volatile uint32_t [0]' [-Warray-bounds=]
  810 |       *reg(SPI_MOSI_DLEN_REG(spi_port)) = (len << 3) - 1;
```
- Bus_SPI crashes at runtime. Although the correct `riscv32-esp-elf-gdb` utility is called, [ESP32 Exception Decoder](https://github.com/me-no-dev/EspExceptionDecoder) is unable to decode the stack trace I'm getting with Arduino.

All this and ESP32-P4 being a little endian chip probably means a dedicated Bus_SPI class for ESP32-P4 is needed.

**Comment by PhonicUK** [comment]
Just seconding my interest on this - I've got some S3 projects I'd love to move over to the P4.

**Comment by tobozo** [comment]
btw there's a MIPI-DSI panel profile for M5Stack Tab5 (ESP32-P4 + ILI9881C) in [M5GFX source tree](https://github.com/m5stack/M5GFX/tree/master/src/lgfx/v1/platforms/esp32p4)

**Reply by tobozo** [reply]
M5Stack Tab5 experimental support is available on [M5GFX_Backport branch](https://github.com/lovyan03/LovyanGFX/tree/M5GFX_Backport).

![image](https://github.com/user-attachments/assets/de41a93c-4eae-494e-b2c8-2589164f60be)

**Comment by tobozo** [comment]
followup for the [Olimex ESP32-P4-DevKit](https://www.olimex.com/Products/IoT/ESP32-P4/) + Bus_SPI + ILI9341:

panel started working with current develop branch ~~after **exclusively selecting pin numbers under 31** for wiring~~

![image](https://github.com/user-attachments/assets/f9b4cbf9-235c-44bd-a351-e0d8a94dc41f)

apparently Bus_SPI is masking pin numbers with `& 31`, which has probably been valid until P4 showed up, but will eventually need to be revisited

here's the config I'm using:

[edit] added touch

```cpp
class Olimex_P4_SPI_ILI9341 : public lgfx::LGFX_Device
{
  lgfx::Panel_ILI9341 _panel_instance;
  lgfx::Bus_SPI       _bus_instance;
  lgfx::Light_PWM     _light_instance;

  public:

    Olimex_P4_SPI_ILI9341()
    {

      {
        auto cfg = _bus_instance.config();
        cfg.freq_write = 80000000;
        cfg.freq_read  = 16000000;
        cfg.pin_mosi = GPIO_NUM_23;
        cfg.pin_miso = GPIO_NUM_22;
        cfg.pin_sclk = GPIO_NUM_5;
        cfg.spi_host = SPI2_HOST;
        cfg.pin_dc   = GPIO_NUM_21;
        cfg.spi_mode = 0 ;
        _bus_instance.config(cfg);
        _panel_instance.bus(&_bus_instance);
      }

      {
        auto cfg = _panel_instance.config();
        cfg.pin_cs  = GPIO_NUM_6;
        cfg.pin_rst = GPIO_NUM_4;
        _panel_instance.config(cfg);
      }


      {
        auto cfg = _touch_instance.config();
        auto buscfg = _bus_instance.config();
        cfg.bus_shared = true;
        cfg.pin_mosi = buscfg.pin_mosi;
        cfg.pin_miso = buscfg.pin_miso;
        cfg.pin_sclk = buscfg.pin_sclk;
        cfg.spi_host = buscfg.spi_host;
        cfg.pin_cs   = GPIO_NUM_20;
        _touch_instance.config(cfg);
        _panel_instance.setTouch(&_touch_instance);
      }

      {
        auto cfg = _light_instance.config();
        cfg.pin_bl = GPIO_NUM_2 ; // Backlight pin
        //cfg.invert = false;
        _light_instance.config(cfg);
        _panel_instance.setLight(&_light_instance);
      }

      setPanel(&_panel_instance);

    }
};
```
